### PR TITLE
Deny setting `.spec.kubernetes.allowPrivilegedContainers` field for Shoot clusters `>=v1.25`

### PR DESCRIPTION
**Pull request by shafeeqes**
**How to categorize this PR?**

/area security
/kind enhancement

**What this PR does / why we need it**:
This PR adds a validation to deny setting `.spec.kubernetes.allowPrivilegedContainers` for Shoot clusters with kubernetes version `>=v1.25`.
With `PodSecurityPolicy` plugin not present, this field doesn't have any relevance.
Users can set their desired level of pod admission by either labelling the namespaces with a [level](https://kubernetes.io/docs/concepts/security/pod-security-admission/#pod-security-levels) or setting a default config in the Shoot spec [Ref](https://github.com/gardener/gardener/blob/master/docs/usage/pod-security.md#admission-configuration-for-the-podsecurity-admission-plugin).

**Which issue(s) this PR fixes**:
Part of #5250

**Special notes for your reviewer**:
~/hold depends on #6549~ 

**Release note**:

```breaking user
For Shoot clusters with kubernetes `v1.25`+ `.spec.kubernetes.allowPrivilegedContainers` should not be set. Please see [here](https://github.com/gardener/gardener/blob/master/docs/usage/pod-security.md#speckubernetesallowprivilegedcontainers-in-the-shoot-spec).
```

### CODE REVIEW

File: pkg/apis/core/types_shoot.go [hunk 1]
```diff
@@ -401,7 +401,8 @@ type HibernationSchedule struct {
 
 // Kubernetes contains the version and configuration variables for the Shoot control plane.
 type Kubernetes struct {
-	// AllowPrivilegedContainers indicates whether privileged containers are allowed in the Shoot (default: true).
+	// AllowPrivilegedContainers indicates whether privileged containers are allowed in the Shoot.
+	// Defaults to true for kubernetes versions below v1.25.
```
Comment (acumino) [inline pkg/apis/core/types_shoot.go]: Nit
Similar changes for other occurrences.
```suggestion
	// AllowPrivilegedContainers indicates whether privileged containers are allowed in the Shoot.
	// Defaults to true for Kubernetes versions below v1.25.
```

**Review by acumino — CHANGES_REQUESTED** [review]

**Review by acumino — COMMENTED** [review]
/lgtm

File: pkg/registry/core/shoot/strategy.go [hunk 2]
```diff
@@ -65,6 +66,9 @@ func (shootStrategy) PrepareForCreate(ctx context.Context, obj runtime.Object) {
 
 	shoot.Generation = 1
 	shoot.Status = core.ShootStatus{}
+
+	// Defaults spec.kubernetes.allowPrivilegedContainers=false for k8s version >= 1.25 and true for lower versions
+	defaultAllowPrivilegedContainers(shoot)
```
Comment (rfranzke) [inline pkg/registry/core/shoot/strategy.go]: Why are we defaulting here in the `registry` package and not in the `defaults.go`?

**Review by rfranzke — COMMENTED** [review]
/assign @rfranzke

File: pkg/registry/core/shoot/strategy.go (hunk 2 above)
Comment (shafeeqes) [reply pkg/registry/core/shoot/strategy.go]: Oops. Somewhere in my mind , It was there that we had to do this only for update calls, so chose not to do it in `defaults.go`. But turns out that's not the case.

**Comment by timebertt** [comment]
/hold
because of some doubts (discussed via Slack)

**Comment by shafeeqes** [comment]
I have reworked the PR as discussed, PTAL! @rfranzke

**Review by acumino — COMMENTED** [review]
We will require to adapt the test machinery test after this PR as most of the test uses the latest Kubernetes version and uses `allowPrivilegedContainers` field.

File: docs/usage/pod-security.md [hunk 3]
```diff
@@ -13,3 +13,43 @@ in `spec.kubernetes.kubeAPIServer.admissionPlugins` field in the `Shoot` resourc
...
+      warn-version: "latest"
+    exemptions:
+      # Array of authenticated usernames to exempt.
+      usernames: []
+      # Array of runtime class names to exempt.
+      runtimeClasses: []
+      # Array of namespaces to exempt.
+      namespaces: []
+```
+For proper functioning of Gardener, if not present, `kube-system` namespace will be added exempted to the exempted namespaces by Gardener.
```
Comment (rfranzke) [inline docs/usage/pod-security.md]: The same docu is already part of the other PR. Should we remove it from here?

File: docs/usage/pod-security.md [hunk 4]
```diff
@@ -13,3 +13,43 @@ in `spec.kubernetes.kubeAPIServer.admissionPlugins` field in the `Shoot` resourc
...
+
+## `.spec.kubernetes.allowPrivilegedContainers` in the Shoot spec
+
+This field if set to `true` means that all authenticated users can use the "gardener.privileged" `PodSecurityPolicy`, allowing full unrestricted access to Pod features. But the `PodSecurityPolicy` admission plugin is removed in Kubernetes `v1.25` and `PodSecurity` has taken its place. Therefore, this field doesn't have any relevance in versions `>=1.25`. If you need to set a default pod admission level for your cluster, follow [Admission Configuration for the PodSecurity admission plugin](#admission-configuration-for-the-podsecurity-admission-plugin).
```
Comment (rfranzke) [inline docs/usage/pod-security.md]:
```suggestion
If this field is set to `true` then all authenticated users can use the "gardener.privileged" `PodSecurityPolicy`, allowing full unrestricted access to Pod features. However, the `PodSecurityPolicy` admission plugin is removed in Kubernetes `v1.25` and `PodSecurity` has is its successor. Therefore, this field doesn't have any relevance in versions `>= v1.25` anymore. If you need to set a default pod admission level for your cluster, follow [this documentation](#admission-configuration-for-the-podsecurity-admission-plugin). 
```

File: pkg/apis/core/v1alpha1/defaults_test.go [hunk 5]
```diff
@@ -632,6 +638,32 @@ var _ = Describe("Defaults", func() {
...
+					obj.Spec.Kubernetes.AllowPrivilegedContainers = pointer.Bool(false)
+				})
+
+				It("should not set the field", func() {
+					SetDefaults_Shoot(obj)
+
+					Expect(obj.Spec.Kubernetes.AllowPrivilegedContainers).To(PointTo(BeFalse()))
+				})
+			})
+		})
```
Comment (rfranzke) [inline pkg/apis/core/v1alpha1/defaults_test.go:666]: What about adding a test for 1.25 version that it's not defaulted (same for `v1beta1`)?

File: pkg/apis/core/v1beta1/types_shoot.go [hunk 6]
```diff
@@ -484,7 +484,8 @@ type HibernationSchedule struct {
 
 // Kubernetes contains the version and configuration variables for the Shoot control plane.
 type Kubernetes struct {
-	// AllowPrivilegedContainers indicates whether privileged containers are allowed in the Shoot (default: true).
+	// AllowPrivilegedContainers indicates whether privileged containers are allowed in the Shoot.
+	// Defaults to true for Kubernetes versions below v1.25.
```
Comment (rfranzke) [inline pkg/apis/core/v1beta1/types_shoot.go]:
```suggestion
	// Defaults to true for Kubernetes versions below v1.25. Unusable for Kubernetes versions v1.25 and higher.
```

(same for other doc strings in types files)

File: pkg/apis/core/validation/shoot.go [hunk 7]
```diff
@@ -867,6 +867,11 @@ func validateKubernetes(kubernetes core.Kubernetes, dockerConfigured, shootHasDe
 		allErrs = append(allErrs, ValidateVerticalPodAutoscaler(*verticalPodAutoscaler, fldPath.Child("verticalPodAutoscaler"))...)
 	}
 
+	k8sGreaterEqual125, _ := versionutils.CheckVersionMeetsConstraint(kubernetes.Version, ">= 1.25")
+	if k8sGreaterEqual125 && kubernetes.AllowPrivilegedContainers != nil {
+		allErrs = append(allErrs, field.Forbidden(fldPath.Child("allowPrivilegedContainers"), "for Kubernetes versions >= 1.25, allowPrivilegedContainers field should not be set, Please see https://github.com/gardener/gardener/blob/master/docs/usage/pod-security.md#speckubernetesallowprivilegedcontainers-in-the-shoot-spec"))
```
Comment (rfranzke) [inline pkg/apis/core/validation/shoot.go]:
```suggestion
		allErrs = append(allErrs, field.Forbidden(fldPath.Child("allowPrivilegedContainers"), "for Kubernetes versions >= 1.25, allowPrivilegedContainers field should not be set, please see https://github.com/gardener/gardener/blob/master/docs/usage/pod-security.md#speckubernetesallowprivilegedcontainers-in-the-shoot-spec"))
```

File: pkg/operation/botanist/kubeapiserver.go [hunk 8]
```diff
@@ -41,11 +41,41 @@ import (
...
+	"k8s.io/apimachinery/pkg/runtime/serializer/json"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
 	"k8s.io/apimachinery/pkg/util/sets"
+	admissionapiv1beta1 "k8s.io/pod-security-admission/admission/api/v1beta1"
 	"k8s.io/utils/pointer"
+	"k8s.io/utils/strings/slices"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
+var (
```
Comment (rfranzke) [inline pkg/operation/botanist/kubeapiserver.go]: Why is all this part of this PR?

File: docs/usage/pod-security.md (hunk 4 above)
Comment (shafeeqes) [reply docs/usage/pod-security.md]: Can you please rephrase, "However, the `PodSecurityPolicy` admission plugin is removed in Kubernetes `v1.25` and `PodSecurity` has is its successor" ? Did you mean, "has taken its place as its successor"?

Comment (rfranzke) [reply docs/usage/pod-security.md]: Sorry:

> However, the `PodSecurityPolicy` admission plugin is removed in Kubernetes `v1.25` and `PodSecurity` is its successor.

But your proposal is also fine for me.

File: pkg/operation/botanist/kubeapiserver.go (hunk 8 above)
Comment (shafeeqes) [reply pkg/operation/botanist/kubeapiserver.go]: Sorry for the confusion, I will rebase this PR once #6549 is merged

**Review by rfranzke — COMMENTED** [review]
/lgtm
/approve

**Comment by shafeeqes** [comment]
> @shafeeqes: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name	Commit	Details	Required	Rerun command
> pull-gardener-integration	[1bef66f](https://github.com/gardener/gardener/commit/1bef66f1d1a2e5933a9872a3b8338724a7b552f0)	[link](https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_gardener/6570/pull-gardener-integration/1565631458125549568)	unknown	`/test pull-gardener-integration`
> [Full PR test history](https://prow.gardener.cloud/pr-history?org=gardener&repo=gardener&pr=6570). [Your PR dashboard](https://prow.gardener.cloud/pr?query=is%3Apr%20state%3Aopen%20author%3Ashafeeqes). [Command help for this repository](https://prow.gardener.cloud/command-help?repo=gardener%2Fgardener). Please help us cut down on flakes by linking this test failure to an [open flake report](https://github.com/gardener/gardener/issues?q=is:issue+is:open+label:kind/flake) or filing a [new flake report](https://github.com/gardener/gardener/issues/new?template=flaking-test.md&title=%5BFlaky+Test%5D+FLAKING+TEST%2FSUITE) if you can't find an existing one. Also see our [testing guideline](https://github.com/gardener/gardener/blob/master/docs/development/testing.md) for how to avoid and hunt flakes.

Failure is due to https://github.com/gardener/gardener/issues/6607
/retest

**Comment by shafeeqes** [comment]
Failures are not related to this PR. 
https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_gardener/6570/pull-gardener-e2e-kind/1565631458054246400
```
Summarizing 2 Failures:
  [FAIL] Shoot Tests [It] Create Shoot, Rotate Credentials and Delete Shoot [Shoot, default, credentials-rotation]
  /home/prow/go/src/github.com/gardener/gardener/test/e2e/shoot/internal/rotation/observability.go:61
  [FAIL] ManagedSeed Tests [It] Create Shoot, Create ManagedSeed, Delete ManagedSeed, Delete Shoot [ManagedSeed, default]
  /home/prow/go/src/github.com/gardener/gardener/test/e2e/managedseed/create_rotate_delete.go:398
```
The first one says 
```
 Expected
      <int>: 503
  to equal
      <int>: 200
  In [It] at: /home/prow/go/src/github.com/gardener/gardener/test/e2e/shoot/internal/rotation/observability.go:61
```
503: Status code is for Service Unavailable. Not sure if it's a flake.

Opened https://github.com/gardener/gardener/issues/6618 for the second one.
/retest

**Pull request merged**